activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1186,6 +1212,14 @@ def check_if_method_has_arguments!(method_name, args)
         end
       end
 
+      def check_if_array_contains_valid_elements!(records)
+        records.flatten.compact.each do |record|
```

Comment:
How likely are users to pass in nested arrays? And ones where `nil` is part of the array?

Does the preloader do the same work on the `records` or does it support nested arrays and nil values?

Since we mutate records how can users properly loop through them if they're either nested or contain nil? E.g. should we punt the `flatten.compact` to user code instead of baking it in here?

Come to think of it, I'm curious if we even need this defensive coding method at all?

Reply:
Preloader handles `nil` with a compact [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/preloader.rb#L86). The [docs](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/preloader.rb#L59-L62) say it should flatten nested arrays but now that I try it, I'm not seeing that behaviour. Will investigate further and see if either docs or code need to be changed.

I wanted to add this layer so that errors don't originate in Preloader, which isn't part of the public API. If we keep flatten, perhaps an addition to documentation (similar to the Preloader doc) would help explain the mutation to `records`.

Reply:
I checked on master and Preloader also doesn't flatten arrays there (i.e. passing a nested array into the `records` parameter causes an `undefined method for Array` error). I'll leave the nil check in, remove the flatten, and update the Preloader docs.